pyccel/errors/messages.py

```diff
@@ -129,3 +129,4 @@
 FOUND_IS_IN_ASSIGN = 'Found `is` statement in assignment [Ignored]'
 ARRAY_REALLOCATION = 'Array redefinition may cause memory reallocation at runtime'
 ARRAY_DEFINITION_IN_LOOP = 'Array definition in for loop may cause memory reallocation at each cycle. Consider creating the array before the loop'
+TEMPLATE_IN_UNIONTYPE = 'Cannot use templates in a union type'
```

Comment:
Since allowing templates inside unions makes the `@types` annotations quite confusing, I totally agree with not allowing it at all!

@EmilyBourne, @saidctb Do you also agree with me and @rmahjoubi?

Reply:
I agree, but I would like a more verbose error message. I am not certain that a user who wasn't involved in the technical discussions would know what you mean by a `union type`. Maybe:
```
Only one header may be provided when using template types
```
?

Reply:
By union we mean `@types('int | float')` I think, so we don't want to allow the following:
```python
@template('T', types=['float', 'complex'])
@types('int | T')
```

Reply:
Ah, that makes even more sense. I didn't realise we'd kept union types. Don't templates do the same job?

Reply:
I also did not realize that at the beginning! :sweat_smile: That's why @rmahjoubi and I had a hard time understanding each other! :rofl: Unions cover only part of what a template can do, because they only allow tensor products.

Reply:
For functions that only require a single template for a single argument, a union is more handy to write. But in all other cases templates are much better, being more explicit and compact at the same time.

Reply:
how about something like this ?
```
@template('T', types=['float', 'complex'])
@types('T[;,:,:] | T[:.:]')
```

Reply:
I cannot imagine many functions that would work for 2D and 3D arrays. Unless they are elemental

Reply:
@rmahjoubi Is the following signature allowed right now?
```python
@template('T', types=['float', 'complex'])
@types('T[:,:]')
```
or does it have to be like this?
```python
@templates('T', types=['float[:,:]', 'complex[:,:]')
@types('T')
```

Reply:
you can easily have one that computes the sum for example , so all vectorized operations + reduction

Reply:
Right now It has to be like this:
```
@template('T', types=['float[:,:]', 'complex[:,:]'])
@types('T')
```

Reply:
> you can easily have one that computes the sum for example , so all vectorized operations + reduction

I you just compute the sum you have two options:
1. Use `numpy.sum()`, hence you don't need Pyccel;
2. Flatten the array first and then use 1D arrays, hence you don't need unions.

Reply:
you are thinking about a simple example, imagine having big code where different functions call each other and you need the sum , so using numpy is not an option, and flattening it also

Reply:
Why is using numpy not an option?